### ISSUE TITLE
RPC for fetching currently set default sock impl

### ISSUE THREAD

**Issue by karlatec** [spdk/spdk#3280]
# Sighting report

When working on https://github.com/spdk/spdk/issues/3090 and running https://github.com/spdk/spdk/blob/11f5dbfa3d34213b6bea011de33ec3ece458b29d/test/nvmf/target/perf_adq.sh#L42 with SPDK built with `--with-uring` option I noticed that `uring` became the defaul sock impl, even though the test script does not change it anywhere.

For development and test writing purposes it might be useful to have `sock_get_default_impl` rpc command. We have a way explicitly set it using `sock_set_default_impl`, but looks like there's no way to `get` the value for this option.

**Comment by KKaras169169** [comment]
We could either set `g_default_impl` during `SPDK_NET_IMPL_REGISTER()` based on current top priority or call `spdk_sock_set_default_impl()` explicitly with `posix` as argument.
A suggestion from @ksztyber is to check for `g_default_impl == NULL` condition at some point and then set the `g_default_impl` to the implementation with top priority.

Either way `sock_get_default_impl` RPC method would require that `g_default_impl` is always set (to match the behavior of `sock_set_default_impl`, which always uses this global variable via `spdk_sock_set_default_impl()`; the new getter would use `spdk_sock_get_default_impl()`). Currently we can check the default implementation only after calling `sock_set_default_impl` and then `save_config`.

**Comment by jimharris** [comment]
> We could either set `g_default_impl` during `SPDK_NET_IMPL_REGISTER()` based on current top priority or call `spdk_sock_set_default_impl()` explicitly with top priority implementation as argument. A suggestion from @ksztyber is to check for `g_default_impl == NULL` condition during GET and return the implementation with top priority if the condition is true.
> 
> Either way `sock_get_default_impl` RPC method would require that `g_default_impl` is always set (to match the behavior of `sock_set_default_impl`, which always uses this global variable via `spdk_sock_set_default_impl()`; the new getter would use `spdk_sock_get_default_impl()`). Currently we can check the default implementation only after calling `sock_set_default_impl` and then `save_config`.

I agree, let's just always set g_default_impl. And change the save_config() function to assert that g_default_impl is set rather than checking it.

Currently uring registers itself with higher priority than posix or ssl. I think we should just make posix the highest priority, which then becomes the default. Adding `--with-uring` shouldn't change the default sock impl, especially since some users (i.e. #3275 submitter) are only enabling it for bdev-related uring reasons.

**Comment by KKaras169169** [comment]
> I agree, let's just always set g_default_impl.

There is also one thing that bothers me here. If I understand the code in `sock.c` correctly, setting `g_default_impl` means that only that single socket implementation is being used to connect (we do not go over all priorities and just return NULL in `spdk_sock_connect_ext()` when we could not get the socket created).

So current behavior lets us try multiple socket impls unless user explicitly requests a single, default implementation. If we always set `g_default_impl` then we loose the ability to try non-default socket implementations.

On the other hand, we could get rid of `g_default_impl` completely and shuffle `g_net_impls` list: the default implementation would be the one with the highest priority (so first on the list) and others could be reassigned a new, lower priority based on their current position in the list. That would unfortunately mean that user cannot request only one socket impl to be used - we would be going over all of them priority-wise.

TL;DR: I think that if we always set `g_default_impl` then socket implementation priority does not make much sense.

**Comment by jimharris** [comment]
> > I agree, let's just always set g_default_impl.
> 
> There is also one thing that bothers me here. If I understand the code in `sock.c` correctly, setting `g_default_impl` means that only that single socket implementation is being used to connect (we do not go over all priorities and just return NULL in `spdk_sock_connect_ext()` when we could not get the socket created).

We can change the sock.c code. Get rid of g_default_impl, always put the "default impl" at the head g_net_impls. If we change the default impl, we can put the old one back into g_net_impls at its proper place by its priority value.

But honestly I'm not sure why we thought it was a good idea to try multiple impls. I think if we say "I want to use uring impl", and some connect or listen fails, we want to know that explicitly, not implicitly try posix (or some other impl).

If there is some good reason to try multiple impls, we should have some way to define that explicitly.

For now, I propose we raise posix priority to make it highest priority. And then change the sock.c code to get rid of g_default_impl as described above. I think this preserves existing behavior. Changing behavior if the default_impl fails (and connect/listen did not specify an impl) can be discussed separately.



> So current behavior lets us try multiple socket impls unless user explicitly requests a single, default implementation. If we always set `g_default_impl` then we loose the ability to try non-default socket implementations.
> 
> On the other hand, we could get rid of `g_default_impl` completely and shuffle `g_net_impls` list: the default implementation would be the one with the highest priority (so first on the list) and others could be reassigned a new, lower priority based on their current position in the list. That would unfortunately mean that user cannot request only one socket impl to be used - we would be going over all of them priority-wise.
> 
> TL;DR: I think that if we always set `g_default_impl` then socket implementation priority does not make much sense.

**Comment by karlowich** [comment]
@jimharris @KKaras169169 I have looked into #3275 and it seems that the problem is that `lib/sock/sock.c/spdk_sock_group_create()` initializes every `impl` in `g_net_impls`. Thus if initialization fails for one of them, you will see an error regardless of what you put as `g_default_impl`.
As far as I can tell this will not be fixed by the patchset above.

**Comment by KKaras169169** [comment]
> the problem is that `lib/sock/sock.c/spdk_sock_group_create()` initializes every `impl` in `g_net_impls`. Thus if initialization fails for one of them, you will see an error regardless of what you put as `g_default_impl`. As far as I can tell this will not be fixed by the patchset above.

Yes, I think the problem here is that we insert all available implementations into `g_net_impls` very early. Then we can only catch initialization errors when we try to allocate sock groups in `spdk_sock_group_create()`. We could make changes so that initialization of sock groups for single sock impl is possible - that would be a workaround sufficient for now, but I think we'd need to try and change how `g_net_impls` is populated or differentiate between 'available' and 'active' implementations in the future.

Also from #3275:

> uring.c:1701:uring_sock_group_impl_create: *ERROR*: Failed to create buffer ring. Your kernel is likely not new enough. Please switch to the POSIX sock implementation instead.

If this is not supposed to be fatal, we could change this log to a warning:

```
WARNING: Failed to create buffer ring for <implementation_name> implementation. Your kernel is likely not new enough.
```

That way it is not as confusing and we can keep application alive after that, because POSIX will have the highest priority.
I could append above changes in the patch [series ](https://review.spdk.io/gerrit/c/spdk/spdk/+/22140) prepared for this issue.

@jimharris @karlowich what do you think?

**Comment by karlowich** [comment]
> If this is not supposed to be fatal, we could change this log to a warning:

I'm guessing this would be fatal if the desired sock implementation is uring.
I think postponing the initialization until an implementation has been selected is the way to go.

Alternatively, if we do like @jimharris proposed:
> Get rid of g_default_impl, always put the "default impl" at the head g_net_impls.

Then perhaps we could initialize only the head?

**Comment by karlowich** [comment]
@jimharris @KKaras169169 Where do we go from here?
Do you need me to do something?

**Comment by jimharris** [comment]
> @jimharris @KKaras169169 Where do we go from here? Do you need me to do something?

I propose we always make 'posix' the default impl, and that we only initialize the selected impl.

I'm not particularly interested in trying to maintain and support the case where if one impl fails, we try some other impl. I feel like that just is asking for problems. I know the code currently is trying to support that, but it seems to just be causing confusion.

I especially don't like that right now if a user wants to use the uring bdev, they have to enable uring when running configure, which then automatically makes uring sock impl the default. That just seems wrong.  Let's make it explicit. I think it's important too as NVIDIA plans to add xlio sock module - ideally we have the xlio support just compiled into SPDK applications by default, but users still need to opt-in to use it.

@KKaras169169 @tomzawadzki @benlwalker what do you think?

**Comment by KKaras169169** [comment]
There is an ongoing development effort in this [patch series](https://review.spdk.io/gerrit/c/spdk/spdk/+/22138)

**Comment by jimharris** [comment]
> There is an ongoing development effort in this [patch series](https://review.spdk.io/gerrit/c/spdk/spdk/+/22138)

Thanks @KKaras169169, I left some comments for you on Gerrit.

**Comment by KKaras169169** [comment]
> @KKaras169169 I think removing `g_default_impl` and using the head of `g_net_impls` instead is the way to go. Then `sock_set_default_impl()` can be used to change the head of `g_net_impls` and `sock_get_default_impl()` can be used to get the head.
> 
> With this in place, we can initialize only the head in `spdk_sock_group_create()` and resolve #3275.
> 
> @jimharris does this align with what we talked about?

Considering [Jim's comments](https://review.spdk.io/gerrit/c/spdk/spdk/+/22690/2//COMMIT_MSG#7) under my patch series, I believe the plan has changed a bit. Socket implementation priorites are a bit confusing, so we are looking into a way to do a few things at once:

1. use POSIX as the default implementation, since it is most likely to work in any given environment due to its simplicity,
2. always clearly and explicitly set the default implementation (instead of what we currently have: a variable to track changes made via RPC and priorities to try and get any socket to work if that is not set),
3. apply the [patch](https://review.spdk.io/gerrit/c/spdk/spdk/+/22140) containing `sock_get_default_impl` on top, after above issues are resolved.

@jimharris correct me if I am mistaken and misinterpreted the suggestions.

**Comment by jimharris** [comment]
I just think these priorities should be configured by the user, and that we don't try to automagically set them.

io_uring is a good example. Let's say we end up with an io_uring implementation that is unequivocally faster and be used over posix, but only for certain kernel versions >= X.Y (this seems plausible). We could make the priority here dependent on kernel version but it definitely complicates how the priorities are determined (in a way we don't support today).

Or for xlio, how does SPDK determine if an interface is xlio capable? I'm sure there are ways to do it, I'm just questioning whether we want SPDK to figure that out itself, or rely on the user to say "use xlio for this listen socket and its accepted connections" and if it doesn't work then fail.

We also currently try all of the impls on listen for example, if xlio fails, try io_uring, if io_uring fails, try posix. But I question whether we should really be doing this kind of fallback. If the interface is supposed to support xlio, and xlio doesn't work, I think we want that to just fail rather than fall back to some other impl.  Same for io_uring.

For the case where we have a heterogenous set of interfaces, we already support creating a listen socket with a user-specified net_impl. So we can have posix as the default, but override it with xlio (or io_uring). So maybe we need to do some work here to expose that through the add_listener RPC.

**Comment by tomzawadzki** [comment]
[Bug scrub] The conclusion from the call is that we should no longer have a priority list at all for difference sock modules. Instead we should only rely on the explicitly provided argument to connect/listen calls. If no option is provided g_default_impl should be used.
`g_default_impl` should remain and be possible to set with RPC, along with new RPC that gets it. 

Relevant changes to iSCSI/NVMe-oF to specify, save and load the impl should be added too.

As for testing changes, either configure option or a better one is fine to add for purpose of testing whole job under specific socket module.
